### PR TITLE
[TetGenInterface] read Smesh as Mesh and move it to MeshLib/IO

### PR DESCRIPTION
**Pull request by norihiro-w**
as titled and see comments in #1164

### CODE REVIEW

File: Applications/DataExplorer/mainwindow.cpp [hunk 1]
```diff
@@ -624,18 +624,22 @@ void MainWindow::loadFile(ImportFileType::type t, const QString &fileName)
     }
     else if (t == ImportFileType::TETGEN)
     {
-        if (fi.suffix().toLower().compare("poly") == 0 || fi.suffix().toLower().compare("smesh") == 0)
+        if (fi.suffix().toLower().compare("smesh") == 0)
```
Comment (TomFischer) [inline Applications/DataExplorer/mainwindow.cpp]: What about poly input format?

Comment (norihiro-w) [inline Applications/DataExplorer/mainwindow.cpp]: i excluded it because They the Tetgen interface is not supporting poly format yet.

**Comment by rinkk** [comment]
I have a number of issues with this PR. I'm going to list these here and maybe we can discuss how to proceed:

* First, a general thing: I don't see the benefit in moving import/export file formats to GeoLib/MeshLib. I can understand that GL/ML should be independent libraries and as such they need to read/write their respective file formats. External formats however are application-specific and the seperation between geometry and mesh is often debatable, as in this case here. If you were to move the OGS-GeoLib into a, say, mechanics code, there'd be no need to read shp-files or GMS-files as these are typical formats for hydrogeological data. If these interfaces are located in FileIO they can work like add-ons for GL/ML and don't clutter up the respective libraries. Also, there are no circular dependencies concerning these interfaces so there is also no reason from a software-engineering POV to move them.

* Concerning this PR: It's debatable if a TetGen - SMesh is a mesh or a geometry file. These files are defining Piecewise Linear Complex (PLC) functions and these could either be viewed as 2D meshes or as triangulated geometric surfaces. Since pre-processing data in OGS is usually geometry and smesh-files *cannot* contain 3D "pieces" I opted to put them into GeoLib. Putting them into MeshLib is the same as putting ArcGIS TIN files into ML or GOCAD TSurfaces. It's certainly possible but our geometric structures represent much better what these files describe.

* Removing the writeSmesh-method for geometries: You removed this but it is actually useful functionality if the input data consists of multiple geometric surfaces, such as GOCAD TSurface files. We do not currently allow to use this in the DataExplorer because it would require stable methods to check if multiple surfaces form a closed volume, but it is usable in scripts/tools where you can assume the user knows what they are doing. In contrast, we do currently not have the functionality to combine multiple 2D meshes into a similar structure.

Sorry for the wall of text.

**Comment by norihiro-w** [comment]
TODO: move tests

**Comment by norihiro-w** [comment]
@rinkk Didn't see #1164 and #1217? Application specific IOs were moved to Applications/FileIO. 
- With current usage, I would keep SMesh in MeshLib because SMesh is indeed with point and topological definitions (i.e. elements). Since TIN doesn't include topological information, it cannot fully represent SMesh, but a mesh can do. 
- If you need Surface object from SMesh, you can do it by using `convertMeshToGeo()` in DE. Opposite conversion is also possible with `convertSurfaceToMesh()`.

If you still don't get it, let's have a short meeting next week. sorry I'm occupied this week.

**Comment by rinkk** [comment]
I am referring to *multiple* surfaces forming *one* geometry and TetGen supporting exactly that.  
Yes, you can convert *one* surface into *one* mesh and vice versa but that doesn't solve the problem of having multiple surface input data from VTK / GOCAD / ArcGIS and converting this.

**Comment by norihiro-w** [comment]
That is something should be done in the conversion. A mesh imported from Smsh has MaterialIDs to distinguish different surfaces.

**Comment by norihiro-w** [comment]
@rinkk I added GeoLib/IO/TetGenInterface for direct IO between Surface objects and a Tetgen smesh file.

**Comment by norihiro-w** [comment]
@rinkk I added also TetGenInterface to GeoLib/IO which can directly make a conversion between Surface objects and Smesh file as it was done previously. I hope this makes you happy

**Comment by norihiro-w** [comment]
can someone review this?

**Comment by norihiro-w** [comment]
modified as discussed in the meeting.

**Comment by endJunction** [comment]
Looks good to me.
:fast_forward:

**Comment by rinkk** [comment]
:+1:

**Pull request merged**